### PR TITLE
Skip variables that don't have dataset or additional_dataset keys

### LINKED ISSUE #843: Error message if no dataset is defined for a variable is very unclear
```
2020-10-22 08:12:15,832 UTC [48962] INFO    esmvalcore._recipe:1298 Creating preprocessor task cmip5/alb
Traceback (most recent call last):
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_main.py", line 430, in run
    fire.Fire(ESMValTool())
  File "/work/bd0854/b309192/soft/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 138, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/work/bd0854/b309192/soft/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 463, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/work/bd0854/b309192/soft/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/fire/core.py", line 672, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_main.py", line 407, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_main.py", line 94, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_recipe.py", line 71, in read_recipe_file
    return Recipe(raw_recipe,
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_recipe.py", line 951, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_recipe.py", line 1299, in initialize_tasks
    task = _get_preprocessor_task(
  File "/mnt/lustre01/pf/b/b309192/ESMValCore/esmvalcore/_recipe.py", line 882, in _get_preprocessor_task
    variable = variables[0]
IndexError: list index out of range
```
from [this](https://github.com/ESMValGroup/ESMValTool/files/5421974/main_log_debug_recipe_preprocessor_derive_test.txt) log running [this](https://github.com/ESMValGroup/ESMValTool/files/5421990/recipe_preprocessor_derive_test.yml.txt) recipe.

### PR DESCRIPTION
**Pull request by valeriupredoi**
This skips it - I can't think of a vulnerability issue with skipping, other than the diagnostic might crash and burn (which is pretty bad!) so do you think we should sys.exit it right there and then? I would argue a skip might be useful if the user wants to run some more diagnostics...

Before you start, please read our [contribution guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [ ] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [ ] If writing a new/modified preprocessor function, please update the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [ ] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #843

### CODE REVIEW

File: esmvalcore/_recipe.py [hunk 1]
```diff
@@ -1252,6 +1252,13 @@ def initialize_tasks(self):
 
             # Create preprocessor tasks
             for variable_group in diagnostic['preprocessor_output']:
+                if not diagnostic['preprocessor_output'][variable_group]:
```
Comment (bouweandela) [inline esmvalcore/_recipe.py:1255]: It's very late in the process of reading the recipe to put this here, a better place would be to do it here:
https://github.com/ESMValGroup/ESMValCore/blob/e1a4a117cb4230b3a4751ec09f6e89fa057e887a/esmvalcore/_recipe.py#L1013
maybe just change the check from checking only for duplicate datasets to also checking for no datasets at all

Comment (valeriupredoi) [reply esmvalcore/_recipe.py:1255]: what a good point! cheers dude!

Comment (valeriupredoi) [reply esmvalcore/_recipe.py]: OK changed it and had to remind meself of the exception handling with Fire :fire:

File: esmvalcore/_recipe.py [hunk 2]
```diff
@@ -1010,6 +1010,10 @@ def _initialize_variables(self, raw_variable, raw_datasets):
         raw_variable = deepcopy(raw_variable)
         datasets = self._initialize_datasets(
             raw_datasets + raw_variable.pop('additional_datasets', []))
+        if not datasets:
+            raise RecipeError(f"You have not specified any dataset "
+                              f"or additional_dataset groups "
+                              f"for variable %s Exiting." % raw_variable)
```
Comment (bouweandela) [inline esmvalcore/_recipe.py:1016]:
```suggestion
            raise RecipeError("You have not specified any dataset "
                              "or additional_dataset groups "
                              f"for variable {raw_variable} Exiting.")
```
Why the sudden change to `%` style string formatting?

**Review by bouweandela — COMMENTED** [review]
Looks good now, would it make sense to add a unit test?

**Comment by valeriupredoi** [comment]
> Looks good now, would it make sense to add a unit test?

yep, was about to but then I go distracted :grin: I'll add one now

File: tests/integration/test_recipe.py [hunk 3]
```diff
@@ -194,6 +194,40 @@ def get_recipe(tempdir, content, cfg):
...
+            scripts: null
+        """)
+    exc_message = \
+        "You have not specified any dataset " + \
+        "or additional_dataset groups for variable " + \
+        "{'preprocessor': 'preprocessor_name', 'project': 'CMIP5'," + \
+        " 'mip': 'Amon', 'exp': 'historical', 'ensemble': 'r1i1p1'" + \
+        ", 'start_year': 1999, 'end_year': 2002, 'variable_group':" + \
+        " 'ta', 'short_name': 'ta', 'diagnostic': " + \
+        "'diagnostic_name'} Exiting."
```
Comment (bouweandela) [inline tests/integration/test_recipe.py:225]:
```suggestion
    exc_message = (
        "You have not specified any dataset "
        "or additional_dataset groups for variable "
        "{'preprocessor': 'preprocessor_name', 'project': 'CMIP5',"
        " 'mip': 'Amon', 'exp': 'historical', 'ensemble': 'r1i1p1'"
        ", 'start_year': 1999, 'end_year': 2002, 'variable_group':"
        " 'ta', 'short_name': 'ta', 'diagnostic': "
        "'diagnostic_name'} Exiting.")
```
Sorry to keep bugging you with string formatting, but can you please avoid using all these backslashes and just use brackets instead?

Comment (valeriupredoi) [reply tests/integration/test_recipe.py]: nay bother man, good point! :beer:

**Comment by bouweandela** [comment]
@valeriupredoi Could you merge `master` into this branch so the tests pass?

**Review by jvegasbsc — APPROVED** [review]

**Pull request merged**